designdocs/provider_registration.md

```diff
@@ -0,0 +1,185 @@
+1. Can provider status be revoked from a provider?
+
+## Goals and Non-Goals
+This document will discuss the ways in which service providers can register with
+and participate in the MRC network.
+
+This document will not discuss the specifics of how providers will process
+service requests from users on the network. This document will also not discuss
+specifics around creating and honoring transactions between users.
+
+## Proposal
+The basic workflow for provider registration is as follows:
+1. User stakes balance to become a provider. This is called a "Registration Fee".
+1. The "Registration Fee" is credited to the Chain Treasury (TBD).
+1. The provider's MSA id is added to storage on chain.
+
+Should a provider choose at any point to stop providing services, it can issue a
+request to remove itself from the provider registry:
+1. The provider issues an announcement that it will stop providing services.
+1. The provider's MSA id is removed from storage.
+
+
+Please note:
+* All names are placeholders and may be changed.
+* Types may change as needed during implementation phase
+* Errors in the extrinsic(s) must have different, reasonably-named error enums
+  for each type of error for ease of debugging.
+
+### Types
+* `ProviderRegistrationParams<T: Config>`, the arguments used to register an announcement.
+  * `provider_msa_id`: `MsaId`
+  * `registration_fee`: `Balance`
+  * `provider_metadata`: `ProviderAnnouncementMetadata`
+* `ProviderRegistrationAnnouncement<T: Config>`, the resource that exists on-chain
+  * `block_number`: `BlockNumber`
+  * `provider_msa_id`: `MsaId`
+  * `provider_metadata`: `ProviderAnnouunucementMetadata`
+* `ProviderUnregistrationParams<T:Config>`, the arguments used to unregister an announcement.
+  * `provider_msa_id`: `MsaId`
+* `ProviderUnregistrationAnnouncement<T:Config>`, the resource that exists on-chain
+  * `provider_msa_id`: `MsaId`
+* `ProviderAnnouncementMetadata`
+  * `name`: `Vec<u8>`
+* `Provider`
+  * `provider_msa_id`: `MsaId`
+  * `name`: `Vec<u8>`
+* `ProviderRegistrationAnnouncementOptions<T:Config>`
+  * `provider_msa_id`:  `Option<MsaId>`, the announcer's MSA id.  Pass None() to get all announcements.
+  * `from_block`: `<T::BlockNumber>`, retrieve messages starting at the given block number (inclusive)
+  * `to_block`: `<T::BlockNumber>`, retrieve messages ending at the given block number (inclusive)
+  * `from_index`: `u32`, starting message index
+  * `page_size`: `usize`, retrieve `page_size` messages at a time, up to configured `T::PageSizeMax`. If 0, return `T::PageSizeMax` results
+
+#### Storage
+* `ProviderRegistry<T: Config>`: `StorageMap<MsaId, Provider>`
+  * Stores registered providers and provides lookup functionality via `MsaId`.
+    Existence in this storage structure implies that a provider's fee has been
+    paid and their registration was otherwise successful.
+* `ProviderRegistrationFees<T: Config>`: `StorageMap<MsaId, Balance>`
```

Comment:
Does successfully registering imply that they paid the fee. Are we thinking about refunding them if the decide not to be a provider anymore?

Reply:
Yes, possible reimbursement was part of the motive for having this storage. But also to provide visibility into fee payments (to my knowledge, the Chain Treasure is an unknown quantity).